### ISSUE TITLE
Can't combine SURROGATE_KEY and IDENTICAL_COPY_FOR_VERSIONED_ENTITY for JPA entities

### ISSUE THREAD

**Issue by sburkard** [jqno/equalsverifier#755]
**Describe the bug**
I can't find a way to successfully verify the [recommended equals and hashcode implementations for JPA entities](https://vladmihalcea.com/how-to-implement-equals-and-hashcode-using-the-jpa-entity-identifier/) for related entities with EqualsVerifier, even with the recommendations on the [EqualsVerifier JPA page](https://jqno.nl/equalsverifier/manual/jpa-entities/) 

**To Reproduce**
Here are two related JPA entities: `Person` and `Tag`. They are related many-to-many. A person can be tagged with 0..* tags and a tag can be applied to 0..* persons. 

Both of them have `equals` and `hashcode` implementations according to [this page](https://vladmihalcea.com/how-to-implement-equals-and-hashcode-using-the-jpa-entity-identifier/) to respect the JPA entity contract (also mentioned on the [EqualsVerifier JPA page](https://jqno.nl/equalsverifier/manual/jpa-entities/). So they only use the `id` field for equality and a constant hashcode. 

````
@Entity
@Getter
@Setter
@NoArgsConstructor
public class Person {
    @Id
    @GeneratedValue(strategy= GenerationType.SEQUENCE)
    private Long id;
    @Column(unique = true)
    private String name;

    @ManyToMany(mappedBy = "persons")
    private Set<Tag> tags;
    @Override
    public int hashCode() {
        return getClass().hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (obj == null)
            return false;
        if (getClass() != obj.getClass())
            return false;

        Person other = (Person) obj;
        return id != null && id.equals(other.getId());
    }
````
````
@Entity
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Tag {
    @Id
    @GeneratedValue(strategy= GenerationType.SEQUENCE)
    private Long id;
    @Column(unique = true)
    private String name;
    @ManyToMany
    private Set<Person> persons;

    @Override
    public int hashCode() {
        return getClass().hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (obj == null)
            return false;
        if (getClass() != obj.getClass())
            return false;

        Tag other = (Tag) obj;
        return id != null && id.equals(other.getId());
    }
}
````

**Code that triggers the behavior**
Now I try to verify `equals` and `hashcode` methods with EqualsVerifier. 

First attempt: Use suppress SURROGATE_KEY according to [EqualsVerifier JPA page](https://jqno.nl/equalsverifier/manual/jpa-entities/) since I only use `id` for equality check
````
EqualsVerifier.simple().forClass(Person.class).suppress(Warning.SURROGATE_KEY).verify();

Output: -> Recursive datastructure.
Add prefab values for one of the following types: Set<Tag>, Tag, Set<Person>, Person.
````

Second attempt: Add prefab values of Tag as recommended
Note: Perhaps I am doing this wrong, I did not find lots of examples. 
````
Tag foo = Tag.builder().id(1L).name("Foo").build();
Tag bar = Tag.builder().id(2L).name("Bar").build();
EqualsVerifier.simple().forClass(Person.class).suppress(Warning.SURROGATE_KEY)
    .withPrefabValues(Tag.class, foo, bar).verify();

Output: -> Reflexivity: object does not equal an identical copy of itself:
If this is intentional, consider suppressing Warning.IDENTICAL_COPY
````

Third attempt: Add suppress IDENTICAL_COPY_FOR_VERSIONED_ENTITY according to output and [EqualsVerifier JPA page](https://jqno.nl/equalsverifier/manual/jpa-entities/) because transient objects (not yet managed) must never be equal.
````
Tag foo = Tag.builder().id(1L).name("Foo").build();
Tag bar = Tag.builder().id(2L).name("Bar").build();
EqualsVerifier.simple().forClass(Person.class).suppress(Warning.SURROGATE_KEY)
    .suppress(Warning.IDENTICAL_COPY_FOR_VERSIONED_ENTITY)
    .withPrefabValues(Tag.class, foo, bar).verify();

Output: Precondition: you can't suppress Warning.IDENTICAL_COPY_FOR_VERSIONED_ENTITY when Warning.SURROGATE_KEY is also suppressed.
````

Alternative attempt: Remove suppress SURROGATE_KEY from third attempt code
````
Output: -> Significant fields: equals relies on id, but hashCode does not.
````

Alternative second attempt: Add suppress STRICT_HASHCODE according to [EqualsVerifier JPA page](https://jqno.nl/equalsverifier/manual/jpa-entities/)
````
Output: -> JPA Entity: direct reference to field tags used in equals instead of getter getTags
````
This last output leads back to suppressing SURROGATE_KEY because it seems to check other fields than `id`

**Expected behavior**
That EqualsVerifier can verify the JPA specific `equals` and `hashcode` methods.
I need SURROGATE_KEY to tell that I only use the `id` for equality check
I need STRICT_HASHCODE to tell that i use constant hashcode
I need IDENTICAL_COPY_FOR_VERSIONED_ENTITY to treat every unmanaged JPA entity as different
But EqualsVerifier does not allow to combine them all.

**Version**
3.12.3

**Additional context**
This verification works fine for a JPA entity like the above but without relations even though I only use `id` for equality and don't suppress SURROGATE_KEY 
````
EqualsVerifier.simple().forClass(MyEntity.class)
    .suppress(Warning.IDENTICAL_COPY_FOR_VERSIONED_ENTITY)
    .suppress(Warning.STRICT_HASHCODE).verify();
````

**Comment by jqno** [comment]
Indeed, you're right. I assumed at one point that if you have a surrogate key, it wouldn't make sense to allow `IDENTICAL_COPY_FOR_VERSIONED_ENTITY` to be suppressed, but looking at your code, it totally does. I'll remove that check.

**Comment by jqno** [comment]
I've just released version 3.13 with a fix.

**Comment by sburkard** [comment]
Perfect, thanks a lot. The following testcase works now

```
@Test
    void givenJpaEntityPerson_whenTestEqualsHashcode_thenRespectJpaContract() {
        TagExample foo = TagExample.builder().id(1L).name("Foo").build();
        TagExample bar = TagExample.builder().id(2L).name("Bar").build();
        EqualsVerifier.simple()
                .forClass(Person.class)
                .suppress(Warning.SURROGATE_KEY)
                .suppress(Warning.IDENTICAL_COPY_FOR_VERSIONED_ENTITY)
                .suppress(Warning.STRICT_HASHCODE)
                .withPrefabValues(TagExample.class, foo, bar)
                .verify();
}
```
